askama_derive/src/input.rs

```diff
@@ -4,41 +4,22 @@ use quote::ToTokens;
 
 use shared::path;
 
-use std::path::{Path, PathBuf};
+use std::path::PathBuf;
 
 use syn;
 
 pub struct TemplateInput<'a> {
     pub ast: &'a syn::DeriveInput,
-    pub source: Source,
     pub print: Print,
     pub escaping: EscapeMode,
-    pub ext: Option<String>,
-    pub parent: Option<&'a syn::Type>,
+    pub parent: Option<String>,
     pub path: PathBuf,
+    source: Source,
 }
 
 impl<'a> TemplateInput<'a> {
     pub fn new(ast: &'a syn::DeriveInput) -> TemplateInput<'a> {
-        let mut meta = None;
-        for attr in &ast.attrs {
-            match attr.interpret_meta() {
-                Some(m) => if m.name() == "template" {
-                    meta = Some(m)
-                },
-                None => {
-                    let mut tokens = TokenStream::new();
-                    attr.to_tokens(&mut tokens);
-                    panic!("unable to interpret attribute: {}", tokens)
-                }
-            }
-        }
-
-        let meta_list = match meta.expect("no attribute 'template' found") {
-            syn::Meta::List(inner) => inner,
-            _ => panic!("attribute 'template' has incorrect type"),
-        };
-
+        let meta_list = meta_list_from(ast);
```

Comment:
The `TemplateInput::new` method seemed way to complicated and contain a lot of stuff so I moved stuff out.

Reply:
Adding more indirection doesn't necessarily make it less complicated. I would prefer adding good comments to indicate the structure and goals, instead.

Reply:
Well, I don't want to start an argument here. So I'll revert this. But I'm a huge fan of functions actually fitting on my screen. (It still doesn't.)